### PR TITLE
Implement mergeWith

### PR DESCRIPTION
**Pull request by luis3m**
Part of #4886

### CODE REVIEW

File: streams/shared/src/main/scala/zio/stream/experimental/ZStream.scala [hunk 1]
```diff
@@ -1622,8 +1622,23 @@ class ZStream[-R, +E, +A](val channel: ZChannel[R, Any, Any, Any, E, Chunk[A], A
...
+    def handler(terminate: Boolean)(exit: Exit[E1, Any]): ZChannel.MergeDecision[R1, E1, Any, E1, Any] =
+      if (terminate || !exit.succeeded) ZChannel.MergeDecision.done(ZIO.done(exit))
+      else ZChannel.MergeDecision.await(ZIO.done(_))
+
+    new ZStream(
+      self
+        .map(l)
+        .channel
+        .mergeWith(that.map(r).channel)(
+          handler(List(Either, Left).contains(strategy)),
```
Comment (iravid) [inline streams/shared/src/main/scala/zio/stream/experimental/ZStream.scala:1637]:
```suggestion
          handler(strategy == Either || strategy == Left),
```

File: streams/shared/src/main/scala/zio/stream/experimental/ZStream.scala [hunk 2]
```diff
@@ -1622,8 +1622,23 @@ class ZStream[-R, +E, +A](val channel: ZChannel[R, Any, Any, Any, E, Chunk[A], A
...
+          handler(List(Either, Right).contains(strategy))
```
Comment (iravid) [inline streams/shared/src/main/scala/zio/stream/experimental/ZStream.scala:1638]:
```suggestion
          handler(strategy == Either || strategy == Right)
```

**Review by iravid — COMMENTED** [review]
Nice and elegant!

File: streams/shared/src/main/scala/zio/stream/experimental/ZStream.scala (hunk 1 above)
Comment (luis3m) [reply streams/shared/src/main/scala/zio/stream/experimental/ZStream.scala]: @iravid done

**Comment by luis3m** [comment]
@iravid addressed suggestions.

**Review by iravid — APPROVED** [review]

**Pull request merged**